src/coreclr/utilcode/sstring.cpp

```diff
@@ -864,27 +862,17 @@ COUNT_T SString::ConvertToUTF8(SString &s) const
         UNREACHABLE();
     }
 
-    // <TODO> @todo: use WC_NO_BEST_FIT_CHARS </TODO>
-    bool  allAscii;
-    DWORD length;
-
-    HRESULT hr = FString::Unicode_Utf8_Length(GetRawUnicode(), & allAscii, & length);
-
-    if (SUCCEEDED(hr))
-    {
-        s.Resize(length, REPRESENTATION_UTF8);
+    COUNT_T length = WszWideCharToMultiByte(CP_UTF8, 0, GetRawUnicode(), GetRawCount()+1, NULL, 0, NULL, NULL);
```

Comment:
I assume that this used `FString::Unicode_Utf8_Length` to avoid some perf problems in `WszWideCharToMultiByte`. Is switching to `WszWideCharToMultiByte` going to introduce performance regression? (I am particularly worried about Windows.)

Reply:
I will check here.

Regardless, in principle I would prefer to defer to system APIs that we can/should assume are optimized. Having these narrowly defined "optimizations" is a non-trivial tax for an already covoluted space like `SString`.

Reply:
I agree that we should get rid of `FString::*`. We may want to replace it with a call to the minipal UTF8 methods for perf reasons like what the original change tried to do. I would expect that minipal UTF8 convertors are going to have significantly better perf than Windows OS WideCharToMultiByte.

> system APIs that we can/should assume are optimized

Historically, Windows OS WideCharToMultiByte has been significantly slower for UTF8 than one would expect from a reasonably optimized implementation.

Reply:
Would static linking [simdutf](https://github.com/simdutf/simdutf) to the VM for such conversions make sense here?

Reply:
> We may want to replace it with a call to the minipal UTF8 methods for perf reasons like what the original change tried to do.

So that I can get behind. There is a small performance regression here. Let me replace this with the minipal APIs. I chose this direction to match the other API. I'd prefer symmetrical API usage in this case as it avoids confusion.

> Would static linking simdutf to the VM for such conversions make sense here?

That is something we can discuss, but it not for this PR.

Reply:
Given the larger regression here, I am not going to convert this to use the new API, which is sad. With this new observation, it seems the minipal API is going to introduce regressions in other places too. There are good changes in this PR I'd like to keep, but not sure the best way to move forward.

Unless there is push back I am inclined keep the code as is and as we continue to remove the PAL revisit the performance observations with the new minipal. Thoughts?

Reply:
I think we should fix the minipal UTF8 perf before removing the FString optimization, so there is no risk perf regression that somebody may need to spend extra time to investigate.

Reply:
https://github.com/dotnet/runtime/pull/85558#issuecomment-1601993367 - we can push the hotpaths to managed, as @jkotas wrote. If that is not possible for certain scenarios; instead of linking to yet another external lib, we can improve the minipal implementation in-place.

The minipal implementation was ported from coreclr-pal, which in turn was based on corefx old managed implenentation. The corefx implementation has since been vectorized, doesn't have gotos-littered code and CFG it produces for GetCount/GetChar etc. is much straightforward. We can try to adapt the minipal implementation to the same plan as the current UTF8.

2016 managed impl., based on which coreclr-plan impl. was initially written: https://github.com/dotnet/coreclr/blob/21f8416ad4204afc18ce315d99baa5f4ada28d9a/src/mscorlib/src/System/Text/UTF8Encoding.cs

present day: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Text/UTF8Encoding.cs

We can run dowm the git history and see the filter the major refactoring commits.

Reply:
> we can push the hotpaths to managed

That seems unlikely in this case. The use of `SString` is litered throughout the system and there are many places where it is going to be unnatural and difficult to call into managed code. We can try jumping out in some cases, but that isn't something I am inclined to do in this PR.

Reply:
My bar for these types of changes is to avoid perf regressions. (Of course, perf improvements are nice - but it is better to evaluate them separately if they come with tradeoffs.)

Neither the existing FString fast path code nor the WideCharToMultiByte implementation built into Windows use any vectorization tricks. It suggests that we should not need vectorization tricks to avoid the perf regression here.

Reply:
The shape of the code in WideCharToMultiByte built into Windows is actually very similar to the corefx implementation and the minipal. It seems that the minipal implementation picked up some cruft alone the way that makes it quite a bit slower.

Reply:
It is interesting that we picked up minipal implementation for Windows + Unix in mono (and Unix in coreclr as before) in .NET 8 and haven't seen any report of regression. Maybe there is some low-hanging alignment issue which can be tweaked via cl switch?

Reply:
> picked up minipal implementation for Windows + Unix in mono

Mono does not get as much perf scrutiny as CoreCLR.

> low-hanging alignment issue

I do not think it is that. 

From a cursory look, I see two problems:
- For smaller lengths, there is a fixed overhead. A simple FString-like loop that deals with small ASCII-only strings into the `minipal_*` methods should fix that.
- For longer lengths, the code of the core look does not look great (e.g. it uses more registers than necessary). For example, it may help to change https://github.com/dotnet/runtime/blob/7b18be52fe191c7961be28621025710d2d6e3b05/src/native/minipal/utf8.c#L1613-L1618 to  
```
                    *pTarget = (unsigned char)ch;
                    *(pTarget + 2) = (unsigned char)(chc);
                    pSrc += 4;
                    *(pTarget + 1) = (unsigned char)(ch >> 16);
                    *(pTarget + 3) = (unsigned char)(chc >> 16);
                    pTarget += 4;
```